task2/task.go

```diff
@@ -0,0 +1,503 @@
+package task2
+
+type Tree23 interface {
+	Insert(k int) *Node
+	Search(k int) *Node
+	SearchMin() *Node
+	Remove(k int) *Node
+	Split(item *Node) *Node
+	Merge(leaf *Node) *Node
+	Redistribute(leaf *Node) *Node
+	Fix(leaf *Node) *Node
+}
+
+type Node struct {
+	size   int
+	key    [3]int
+	first  *Node
+	second *Node
+	third  *Node
+	fourth *Node
+	parent *Node
+}
+
+func NewNode(k int) *Node {
+	return &Node{
+		size:   1,
+		key:    [3]int{k, 0, 0},
+		first:  nil,
+		second: nil,
+		third:  nil,
+		fourth: nil,
+		parent: nil,
+	}
+}
+
+func (n *Node) find(k int) bool {
+	for i := 0; i < n.size; i++ {
+		if n.key[i] == k {
+			return true
+		}
+	}
+	return false
+}
+
+func (n *Node) swap(x, y *int) {
+	*x, *y = *y, *x
+}
+
+func (n *Node) sort() {
+	if n.size == 1 {
+		return
+	}
+	if n.size == 2 {
+		if n.key[0] > n.key[1] {
+			n.swap(&n.key[0], &n.key[1])
+		}
+	}
+	if n.size == 3 {
+		if n.key[0] > n.key[1] {
+			n.swap(&n.key[0], &n.key[1])
+		}
+		if n.key[0] > n.key[2] {
+			n.swap(&n.key[0], &n.key[2])
+		}
+		if n.key[1] > n.key[2] {
+			n.swap(&n.key[1], &n.key[2])
+		}
+	}
+}
+
+func (n *Node) insertToNode(k int) {
+	n.key[n.size] = k
+	n.size++
+	n.sort()
+}
+
+func (n *Node) removeFromNode(k int) {
+	if n.size >= 1 && n.key[0] == k {
+		n.key[0] = n.key[1]
+		n.key[1] = n.key[2]
+		n.size--
+	} else if n.size == 2 && n.key[1] == k {
+		n.key[1] = n.key[2]
+		n.size--
+	}
+}
+
+func (n *Node) becomeNode2(k int, first, second *Node) {
+	n.key[0] = k
+	n.first = first
+	n.second = second
+	n.third = nil
+	n.fourth = nil
+	n.parent = nil
+	n.size = 1
+}
+
+func (n *Node) isLeaf() bool {
+	return n.first == nil && n.second == nil && n.third == nil
+}
+
+func (n *Node) Insert(k int) *Node {
+	if n == nil {
+		return NewNode(k)
+	}
+
+	if n.isLeaf() {
+		n.insertToNode(k)
+	} else if k <= n.key[0] {
+		n.first = n.first.Insert(k)
+	} else if n.size == 1 || (n.size == 2 && k <= n.key[1]) {
+		n.second = n.second.Insert(k)
+	} else {
+		n.third = n.third.Insert(k)
+	}
```

Comment:
вот тут у тебя багуля, ты присваиваешь потомку его родителя, из-за этого происходит зацикливание в дереве, чего допускать нельзя)
```suggestion
	} else if k <= n.key[0] {
		n.first.Insert(k)
	} else if n.size == 1 || (n.size == 2 && k <= n.key[1]) {
		n.second.Insert(k)
	} else {
		n.third.Insert(k)
	}
```